onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1923,6 +1923,179 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+        // Output type is index type
+        propagateElemTypeFromDtypeToOutput(ctx, TensorProto_DataType_INT64, 0);
+        // Output shape is runtime dependent and can not be inferred as it depends on
+        // the input sparse tensor COO indices shape.
+      }) ;
+
+    static const char* SparseGemm_doc = R"DOC(
+  Behaves as to https://github.com/onnx/onnx/blob/master/docs/Operators.md#Gemm
+  XXX: can we do without transpose now?
+  XXX: It has a lot of common code with SparseToDenseMatMul which we can reuse.
+  Should we merge it?
+)DOC";
+  ONNX_CONTRIB_OPERATOR_SCHEMA(Gemm)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .Input(0, "A", "Sparse matrix (M,K) A COO format", "T")
+      .Input(1, "B", "Sparse matrix (K,N) ", "T")
+      .Input(2, "C", "1 or 2 - dimensional sparse matrix B", "T", ONNX_NAMESPACE::OpSchema::Optional)
+      .Output(0, "Y", "Matrix multiply results", "T")
+      .Attr(
+          "alpha",
+          "Scalar multiplier for the product of the input tensors A * B.",
+          AttributeProto::FLOAT,
+          1.0f)
+        .Attr(
+            "beta",
+            "Scalar multiplier for the product of the input tensor C.",
+            AttributeProto::FLOAT,
+            .0f)
+      .Attr(
+          "transA",
+          "Whether A should be transposed on the last two dimensions before doing multiplication",
+          AttributeProto::INT,
+          static_cast<int64_t>(0))
+      .Attr(
+          "transB",
+          "Whether B should be transposed on the last two dimensions before doing multiplication",
+          AttributeProto::INT,
+          static_cast<int64_t>(0))
+      .Output(0, "Y", "Matrix multiply results", "T")
+      .TypeConstraint("T", {"sparse_tensor(float)"}, "Constrain input and output float at this time")
+      .TypeAndShapeInferenceFunction([](ONNX_NAMESPACE::InferenceContext& ctx) {
+          const auto* input_type0 = ctx.getInputType(0);
+          const auto* input_type1 = ctx.getInputType(1);
+          if (nullptr == input_type0 || nullptr == input_type1) {
+            fail_type_inference("At least on of the inputs missing a type");
+          }
+          if (getTensorElementType(*input_type0) != getTensorElementType(*input_type1)) {
+            fail_type_inference("Both inputs must have the same element type");
+          }
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          // TODO: replace with ONNX one when that one is fixed
+          sparseCompatibleMatmulShapeInference(ctx, 0, 1);
+      });
+
+    static const char* Sum_doc = R"DOC(
+  Behaves as to https://github.com/onnx/onnx/blob/master/docs/Operators.md#Sum
+  For 1-D input COO index format it would output flat 1-D index and for 
+  2-D input COO index format it would output 2-D COO index.
+  Both inputs must have the same dense shape and is the output
```

Comment:
> Both inputs must have the same dense shape and is the output [](http://example.com/codeflow?start=1&length=61)

Is this a reasonable requirement?

Reply:
Our input Sparse Tensors will not have the same dense shape. The example from the word doc:

> Sparse Input 1: 
> indices = value 
> [1001] = 5.0 
> [1002] = 6.0 
> 
> Sparse Input 2: 
> indices = value 
> [20] = 2.0 
> [1002] = 3.0 
> 
> Sparse Output: 
> indices = value 
> [20] = 2.0 
> [1001] = 5.0 
> [1002] = 9.0